### PR TITLE
Implement generics related documentation rules

### LINKED ISSUE #138: SA1618: GenericTypeParametersMustBeDocumented
http://www.stylecop.com/docs/SA1618.html

### LINKED ISSUE #139: SA1619: GenericTypeParametersMustBeDocumentedPartialClass
http://www.stylecop.com/docs/SA1619.html

### LINKED ISSUE #140: SA1620: GenericTypeParameterDocumentationMustMatchTypeParameters
http://www.stylecop.com/docs/SA1620.html

### LINKED ISSUE #141: SA1621: GenericTypeParameterDocumentationMustDeclareParameterName
http://www.stylecop.com/docs/SA1621.html

### LINKED ISSUE #142: SA1622: GenericTypeParameterDocumentationMustHaveText
http://www.stylecop.com/docs/SA1622.html

### PR DESCRIPTION
**Pull request by pdelvo**
This implements all documentation rules for generics

Fixes #138.
Fixes #139.
Fixes #140.
Fixes #141.
Fixes #142.

### CODE REVIEW

File: StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1620GenericTypeParameterDocumentationMustMatchTypeParameters.cs [hunk 1]
```diff
@@ -28,13 +34,16 @@ public class SA1620GenericTypeParameterDocumentationMustMatchTypeParameters : Di
         /// </summary>
         public const string DiagnosticId = "SA1620";
         private const string Title = "Generic type parameter documentation must match type parameters";
-        private const string MessageFormat = "TODO: Message format";
+        private const string MessageFormat = "{0}";
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1620GenericTypeParameterDocumentationMustMatchTypeParameters.cs]: You can create two `DiagnosticDescriptor` instances that have the same ID but different message formats. Only return one of them from `SupportedDiagnostics` but the other can still be used when the second message is needed.

File: StyleCop.Analyzers/StyleCop.Analyzers/Helpers/XmlCommentHelper.cs [hunk 2]
```diff
@@ -18,8 +18,9 @@ internal static class XmlCommentHelper
         internal const string InheritdocXmlTag = "inheritdoc";
         internal const string ReturnsXmlTag = "returns";
         internal const string ValueXmlTag = "value";
-        internal const string SeeXmlTag = "see";
-        internal const string ParamTag = "param";
+        internal const string SeeTag = "see";
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/Helpers/XmlCommentHelper.cs]: :question: Why did you change `SeeXmlTag` &rarr; `SeeTag`, but `ParamTag` &rarr; `ParamXmlTag`?

Comment (pdelvo) [inline StyleCop.Analyzers/StyleCop.Analyzers/Helpers/XmlCommentHelper.cs]: This was by accident. I change it so that every xml tag includes the "xml"

File: StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1620GenericTypeParameterDocumentationMustMatchTypeParameters.cs (hunk 1 above)
Comment (pdelvo) [inline StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1620GenericTypeParameterDocumentationMustMatchTypeParameters.cs]: Okay I will change that

File: StyleCop.Analyzers/StyleCop.Analyzers.Test/DocumentationRules/SA1618UnitTests.cs [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+    using Microsoft.CodeAnalysis.Diagnostics;
+    using TestHelper;
+    using Xunit;
+
+    /// <summary>
+    /// This class contains unit tests for <see cref="SA1618GenericTypeParametersMustBeDocumented"/>.
+    /// </summary>
+    public class SA1618UnitTests : CodeFixVerifier
+    {
+        public string DiagnosticId { get; } = SA1618GenericTypeParametersMustBeDocumented.DiagnosticId;
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers.Test/DocumentationRules/SA1618UnitTests.cs]: :bulb: This isn't used anymore. We used these properties prior to #526.

File: StyleCop.Analyzers/StyleCop.Analyzers.Test/DocumentationRules/SA1618UnitTests.cs [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+    using System.Threading.Tasks;
+    using Analyzers.DocumentationRules;
+    using Microsoft.CodeAnalysis.Diagnostics;
+    using TestHelper;
+    using Xunit;
+
+    /// <summary>
+    /// This class contains unit tests for <see cref="SA1618GenericTypeParametersMustBeDocumented"/>.
+    /// </summary>
+    public class SA1618UnitTests : CodeFixVerifier
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers.Test/DocumentationRules/SA1618UnitTests.cs]: :bulb: Test suites that don't include code fixes should extend `DiagnosticVerifier` instead of `CodeFixVerifier`.

File: StyleCop.Analyzers/StyleCop.Analyzers.Test/DocumentationRules/SA1620UnitTests.cs [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+                yield return new object[] { "class     Foo<Ta, Tb> { }" };
+                yield return new object[] { "struct    Foo<Ta, Tb> { }" };
+                yield return new object[] { "interface Foo<Ta, Tb> { }" };
+                yield return new object[] { "class     Foo<Ta, T\\u0062> { }" };
+                yield return new object[] { "struct    Foo<Ta, T\\u0062> { }" };
+                yield return new object[] { "interface Foo<Ta, T\\u0062> { }" };
+            }
+        }
+
+        public static IEnumerable<object[]> Declarations
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers.Test/DocumentationRules/SA1620UnitTests.cs]: :question: Was this supposed to be used?

File: StyleCop.Analyzers/StyleCop.Analyzers.Test/DocumentationRules/SA1622UnitTests.cs [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+    using System.Collections.Generic;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Analyzers.DocumentationRules;
+    using Microsoft.CodeAnalysis.Diagnostics;
+    using TestHelper;
+    using Xunit;
+
+    /// <summary>
+    /// This class contains unit tests for <see cref="SA1622GenericTypeParameterDocumentationMustHaveText"/>-
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers.Test/DocumentationRules/SA1622UnitTests.cs]: :bulb: This should probably end with `>.` instead of `>-`.

File: StyleCop.Analyzers/StyleCop.Analyzers.Test/DocumentationRules/SA1620UnitTests.cs (hunk 5 above)
Comment (pdelvo) [inline StyleCop.Analyzers/StyleCop.Analyzers.Test/DocumentationRules/SA1620UnitTests.cs]: No it wasnt. I forgot to remove it

File: StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1618GenericTypeParametersMustBeDocumented.cs [hunk 7]
```diff
@@ -48,7 +53,75 @@ public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
...
+            }
+
+            var xmlParameterNames = documentation.Content.GetXmlElements(XmlCommentHelper.TypeParamXmlTag)
+                .Select(XmlCommentHelper.GetFirstAttributeOrDefault<XmlNameAttributeSyntax>)
+                .Where(x => x != null)
+                .ToImmutableArray();
+
+            foreach (var parameter in typeParameterList.Parameters)
+            {
+                if (!xmlParameterNames.Any(x => x.Identifier.Identifier.ValueText == parameter.Identifier.ValueText))
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1618GenericTypeParametersMustBeDocumented.cs]: :question: What happens here when you have syntax like this?

```xml
<typeparam name=""></typeparam>
```

Comment (pdelvo) [inline StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1618GenericTypeParametersMustBeDocumented.cs]: SA1621 and SA1622 are reported, so this works as expected

File: StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1620GenericTypeParameterDocumentationMustMatchTypeParameters.cs [hunk 8]
```diff
@@ -51,7 +60,89 @@ public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
...
+            if (string.IsNullOrWhiteSpace(nameAttribute?.Identifier?.Identifier.ValueText))
+            {
+                return;
+            }
+
+            if (!parentTypeParameters.Contains(nameAttribute.Identifier.Identifier.ValueText))
+            {
+                context.ReportDiagnostic(Diagnostic.Create(Descriptor, nameAttribute?.Identifier?.GetLocation() ?? alternativeDiagnosticLocation, string.Format(MissingTypeParamForDocumentation, nameAttribute.Identifier.Identifier.ValueText)));
+            }
+            else if (parentTypeParameters[index] != nameAttribute.Identifier.Identifier.ValueText)
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1620GenericTypeParameterDocumentationMustMatchTypeParameters.cs]: :exclamation: This goes :boom: if you have more `<typeparam>` elements than generic type parameters.

File: StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1621GenericTypeParameterDocumentationMustDeclareParameterName.cs [hunk 9]
```diff
@@ -50,7 +54,39 @@ public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
...
         public override void Initialize(AnalysisContext context)
         {
-            // TODO: Implement analysis
+            context.RegisterSyntaxNodeActionHonorExclusions(this.HandleXmlElement, SyntaxKind.XmlElement);
+            context.RegisterSyntaxNodeActionHonorExclusions(this.HandleXmlEmptyElement, SyntaxKind.XmlEmptyElement);
+        }
+
+        private void HandleXmlElement(SyntaxNodeAnalysisContext context)
+        {
+            XmlElementSyntax emptyElement = context.Node as XmlElementSyntax;
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1621GenericTypeParameterDocumentationMustDeclareParameterName.cs]: :bulb: `emptyElement` is probably not the best name.

File: StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1621GenericTypeParameterDocumentationMustDeclareParameterName.cs [hunk 10]
```diff
@@ -50,7 +54,39 @@ public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
...
+
+            var name = emptyElement?.StartTag?.Name;
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1621GenericTypeParameterDocumentationMustDeclareParameterName.cs]: :bulb: `emptyElement` cannot be null here.

File: StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1621GenericTypeParameterDocumentationMustDeclareParameterName.cs [hunk 11]
```diff
@@ -50,7 +54,39 @@ public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
...
+
+            HandleElement(context, emptyElement, name, emptyElement?.StartTag?.GetLocation());
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1621GenericTypeParameterDocumentationMustDeclareParameterName.cs]: :bulb: `emptyElement` cannot be null here.

File: StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1621GenericTypeParameterDocumentationMustDeclareParameterName.cs [hunk 12]
```diff
@@ -50,7 +54,39 @@ public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
...
+        }
+
+        private void HandleXmlEmptyElement(SyntaxNodeAnalysisContext context)
+        {
+            XmlEmptyElementSyntax emptyElement = context.Node as XmlEmptyElementSyntax;
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1621GenericTypeParameterDocumentationMustDeclareParameterName.cs]: :bulb: use a cast instead of `as`.

File: StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1621GenericTypeParameterDocumentationMustDeclareParameterName.cs [hunk 13]
```diff
@@ -50,7 +54,39 @@ public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
...
+
+            var name = emptyElement?.Name;
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1621GenericTypeParameterDocumentationMustDeclareParameterName.cs]: :bulb: `emptyElement` cannot be null here.

File: StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1621GenericTypeParameterDocumentationMustDeclareParameterName.cs [hunk 14]
```diff
@@ -50,7 +54,39 @@ public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
...
+
+            HandleElement(context, emptyElement, name, emptyElement?.GetLocation());
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1621GenericTypeParameterDocumentationMustDeclareParameterName.cs]: :bulb: `emptyElement` cannot be null here.
